In [146]:
import pandas as pd
import matplotlib.pyplot as plt

In [147]:
shooters = pd.read_csv('Data/Stanford_MSA_Database.csv')

OK, so there's lots of things we can do with the dataset given that it's a full 335 rows with 55 columns of descriptive text data each. Let's start by coming up with some basic questions we want answered:
- What kinds of weapons are most popular, and where?
- Where are guns most readily accessible in the country?
- How do warning signs change with age and other factors?
- Does lethality change with age?
- Have shootings become, on the whole, more lethal over time?
- How do the number of fatalities compare to the total number of victims?
- Has the number of guns used in a shooting changed on average over time?
- What happens to the shooter, and how does that change with age and race? Does their fate change substantially in incidences of school shootings?
- How do the motivations of mass shooters differ with geography?
- Does military experience alter the motive?

In [148]:
shooters

CaseID  \
0         1   
1         2   
2         3   
3         4   
4         5   
5         6   
6         7   
7         8   
8         9   
9        10   
10       11   
11       12   
12       13   
13       14   
14       15   
15       16   
16       17   
17       18   
18       19   
19       20   
20       21   
21       22   
22       23   
23       24   
24       25   
25       26   
26       27   
27       28   
28       29   
29       30   
..      ...   
305     315   
306     316   
307     317   
308     318   
309     319   
310     320   
311     321   
312     322   
313     323   
314     324   
315     325   
316     326   
317     327   
318     328   
319     329   
320     330   
321     331   
322     332   
323     333   
324     334   
325     335   
326     336   
327     337   
328     338   
329     339   
330     340   
331     341   
332     342   
333     343   
334     344   

                                                                     Title  \
0                                            University of Texas at Austin   
1                                               Rose-Mar College of Beauty   
2                                             New Orleans Police Shootings   
3                                           Clara Barton Elementary School   
4                                                        Olean High School   
5                                     Los Angeles Computer Learning Center   
6                                                      Cal State Fullerton   
7                                       Grover Cleveland Elementary School   
8                                             University of South Carolina   
9                                                       Valley High School   
10                                                   Welding shop in Miami   
11                                                 Wah Mee Club in Seattle   
12                                                    Johnston Post Office   
13                                           49th Street Elementary School   
14                                                  Ianni's Club in Dallas   
15                                     McDonald's restaurant in San Ysidro   
16                                              Goddard Junior High School   
17                                                     Atlanta Post Office   
18                                              New York Technical College   
19                                         Post office in Edmond, Oklahoma   
20                                               Fergus County High School   
21                                               Palm Bay Shopping Centers   
22                                               Pinellas Park High School   
23                                Electromagnetic Systems Laboratory (ESL)   
24                                         Hubbard Woods Elementary School   
25                                  Comet Auto Parts and Montefiore School   
26                                               Oakland Elementary School   
27                                        New Orleans Downtown Post Office   
28                                             Cleveland Elementary School   
29                                                Orange Glenn Post Office   
..                                                                     ...   
305                                         Forestville, Maryland Drive-by   
306                                                  Ferguson, MO Drive by   
307                                              Appling, GA family murder   
308                                               Auburn, AL club shootout   
309                                  Nashville Music City Central shooting   
310                                        Murfreesboro, TN party shooting   
311                                  Florida Ex-wife, ex-girlfriend murder   
312                             

In [149]:
list(shooters)

['CaseID',
 'Title',
 'Location',
 'City',
 'State',
 'Latitude',
 'Longitude',
 'Number of Civilian Fatalities',
 'Number of Civilian Injured',
 'Number of Enforcement Fatalities',
 'Number of Enforcement Injured',
 'Total Number of Fatalities',
 'Total Number of Victims',
 'Description',
 'Date',
 'Day of Week',
 'Date - Detailed',
 'Shooter Name',
 'Number of shooters',
 'Shooter Age(s)',
 'Average Shooter Age',
 'Shooter Sex',
 'Shooter Race',
 'Type of Gun - Detailed',
 'Type of Gun - General',
 'Number of Shotguns',
 'Number of Rifles',
 'Number of Handguns',
 'Total Number of Guns',
 'Number of Automatic Guns',
 'Number of Semi-Automatic Guns',
 'Fate of Shooter at the scene',
 'Fate of Shooter',
 "Shooter's Cause of Death",
 'School Related',
 'Place Type',
 'Relationship to Incident Location',
 'Targeted Victim/s - Detailed',
 'Targeted Victim/s - General',
 'Possible Motive - Detailed',
 'Possible Motive - General',
 'History of Mental Illness - Detailed',
 'History of Mental

I'm curious what's going to happen if I run a market basket analysis. Let's clean up the data a little more and see what happens. I'm working off of the following tutorial:
http://pbpython.com/market-basket-analysis.html

In [150]:
cleaned_shooters = shooters.copy(deep=True)

#Clean gun types
cleaned_shooters['Type of Gun - General'] = cleaned_shooters['Type of Gun - General'].str.strip()
cleaned_shooters['Type of Gun - General'] = cleaned_shooters['Type of Gun - General'].str.lower()
cleaned_shooters['Type of Gun - General'].unique()

array(['multiple guns', 'handgun', 'shotgun', 'rifle', 'unknown',
       'semi-automatic rifle', '9-mm'], dtype=object)

In [151]:
#Clean gun counts
guncount_types = ['Number of Shotguns','Number of Handguns','Number of Rifles']
for guncount_type in guncount_types:
    cleaned_shooters[guncount_type].replace('Unknown', 0, inplace=True)

In [152]:
#Clean military experience
cleaned_shooters['Military Experience'] = cleaned_shooters['Military Experience'].str.lower()
cleaned_shooters['Military Experience'].replace('yes', True, inplace=True)
cleaned_shooters['Military Experience'].replace(['no', 'unknown'], False, inplace=True)

In [153]:
#Set civilian + law enforcement casualties, injuries to bools
cleaned_shooters['Number of Civilian Fatalities'] = cleaned_shooters['Number of Civilian Fatalities'] > 0
cleaned_shooters['Number of Civilian Injured'] = cleaned_shooters['Number of Civilian Injured'] > 0
cleaned_shooters['Number of Enforcement Fatalities'] = cleaned_shooters['Number of Enforcement Fatalities'] > 0
cleaned_shooters['Number of Enforcement Injured'] = cleaned_shooters['Number of Enforcement Injured'] > 0
cleaned_shooters['High Civilian Casualties'] = (shooters['Number of Civilian Fatalities'] + shooters['Number of Civilian Injured']) > 10

In [154]:
#Corner cases (from errors)
cleaned_shooters['Number of Shotguns'].replace('Handgun', 0, inplace=True)
cleaned_shooters['Number of Rifles'].replace('0 (1)', 0, inplace=True)
cleaned_shooters['Number of Handguns'].replace('2 (1)', 0, inplace=True)

In [155]:
#Set gun types to bools
cleaned_shooters['Number of Shotguns'] = pd.to_numeric(cleaned_shooters['Number of Shotguns']) > 0
cleaned_shooters['Number of Rifles'] = pd.to_numeric(cleaned_shooters['Number of Rifles']) > 0
cleaned_shooters['Number of Handguns'] = pd.to_numeric(cleaned_shooters['Number of Handguns']) > 0

In [156]:
#Clean fate of shooter results
cleaned_shooters['Fate of Shooter'].replace('Custody / Escaped', 'Custody/Escaped', inplace=True)

In [157]:
len(cleaned_shooters.loc[
    (cleaned_shooters['Fate of Shooter'].isnull()) | 
    (cleaned_shooters['Fate of Shooter'] == 'FALSE')
])

3

In [158]:
#Drop the three where we don't know the fate of the shooter
cleaned_shooters = cleaned_shooters.loc[
    (~cleaned_shooters['Fate of Shooter'].isnull()) & 
    ~(cleaned_shooters['Fate of Shooter'] == 'FALSE')
]

In [159]:
#We'll iterate over our desired features and flatten them s.t.
#each unique value has a column, and if the incident has that value
#we'll just assign the matching column a 1 and the other ones a 0
marketbasket = cleaned_shooters[[
    'Number of Civilian Fatalities',
    'Number of Civilian Injured',
    'Number of Enforcement Fatalities',
    'Number of Enforcement Injured',
    'Military Experience',
    'Number of Shotguns',
    'Number of Rifles',
    'Number of Handguns',
    'High Civilian Casualties'
]]


#Get the logical inverses as well
marketbasket['No Civilian Fatalities'] = ~marketbasket['Number of Civilian Fatalities']
marketbasket['No Civilian Injured'] = ~marketbasket['Number of Civilian Injured']
marketbasket['No Enforcement Fatalities'] = ~marketbasket['Number of Enforcement Fatalities']
marketbasket['No Enforcement Injured'] = ~marketbasket['Number of Enforcement Injured']


##TODO: Need to finish generating the rest of the dataframe by expanding out unique values across columns
desired_features = [
    'Fate of Shooter'
]

for feature in desired_features:
    for val in cleaned_shooters[feature].unique():
        marketbasket[feature + ' ' + val] = (cleaned_shooters[feature] == val)

marketbasket

C:\Users\spenc\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
C:\Users\spenc\Anaconda3\lib\site-packages\ipykernel_launcher.py:19: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
C:\Users\spenc\Anaconda3\lib\site-packages\ipykernel_launcher.py:20: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexi

Number of Civilian Fatalities  Number of Civilian Injured  \
0                             True                        True   
1                             True                        True   
2                             True                        True   
3                             True                        True   
4                             True                        True   
5                             True                        True   
6                             True                        True   
7                             True                        True   
8                             True                        True   
9                             True                        True   
10                            True                        True   
11                            True                        True   
12                            True                        True   
13                            True                        True   
14                            True                        True   
15                            True                        True   
16                            True                        True   
17                            True                        True   
18                            True                        True   
19                            True                        True   
20                            True                        True   
21                            True                        True   
22                            True                        True   
23                            True                        True   
24                            True                        True   
25                            True                        True   
26                            True                        True   
27                           False                        True   
28                            True                        True   
29                            True                        True   
..                             ...                         ...   
305                           True                        True   
306                          False                        True   
307                           True                       False   
308                           True                        True   
309                          False                        True   
310                           True                       False   
311                           True                       False   
312                           True                        True   
313                           True                        True   
314                           True                        True   
315                           True                        True   
316                           True                        True   
317                           True                        True   
318                           True                        True   
319                           True                        True   
320                          False                        True   
321                           True                       False   
322                           True                       False   
323                           True                       False   
324                           True                        True   
325                           True                        True   
326                           True                        True   
327                           True                        True   
328                          False                        True   
329                           True                        True   
330                          False                        True   
331                           True                        True   
332                          False                        True   
333                           True           

In [160]:
#Apply apriori market basket algorithm
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules
apriori_res = apriori(marketbasket, min_support=0.1, use_colnames=True)

In [161]:
#Create display copy of apriori results
apriori_res_display = apriori_res.copy(deep=True)
apriori_res_display['itemsets'] = apriori_res_display['itemsets'].apply(lambda x: ''.join(e + ', ' for e in x))

In [162]:
pd.options.display.max_colwidth = 300
apriori_res_display.sort_values('support')

support  \
120  0.102410   
185  0.102410   
149  0.102410   
49   0.102410   
84   0.102410   
13   0.102410   
141  0.102410   
138  0.102410   
115  0.102410   
58   0.102410   
201  0.102410   
137  0.102410   
61   0.102410   
116  0.102410   
118  0.102410   
119  0.102410   
208  0.102410   
200  0.102410   
106  0.102410   
123  0.105422   
159  0.105422   
160  0.105422   
98   0.105422   
95   0.105422   
184  0.105422   
205  0.105422   
93   0.108434   
94   0.108434   
155  0.108434   
157  0.108434   
..        ...   
60   0.424699   
101  0.433735   
57   0.433735   
11   0.433735   
100  0.442771   
12   0.445783   
161  0.451807   
27   0.463855   
79   0.466867   
78   0.469880   
18   0.493976   
125  0.524096   
44   0.542169   
43   0.548193   
5    0.575301   
150  0.599398   
67   0.614458   
66   0.626506   
14   0.650602   
109  0.743976   
31   0.759036   
88   0.771084   
30   0.780120   
22   0.789157   
21   0.804217   
1    0.807229   
0    0.834337   
55   0.915663   
10   0.936747   
9    0.960843   

                                                                                                                                                         itemsets  
120                                                                                           Number of Rifles, No Enforcement Injured, Fate of Shooter Custody,   
185                                                                  Number of Handguns, No Civilian Injured, No Enforcement Fatalities, No Enforcement Injured,   
149                                               Number of Civilian Fatalities, Number of Civilian Injured, High Civilian Casualties, Fate of Shooter Deceased,   
49                                                                                                           High Civilian Casualties, Fate of Shooter Deceased,   
84                                                                            Number of Civilian Fatalities, High Civilian Casualties, Fate of Shooter Deceased,   
13                                                                                                                                      Fate of Shooter Escaped,   
141                                                       Number of Civilian Fatalities, Number of Civilian Injured, Number of Rifles, Fate of Shooter Deceased,   
138                                                       Number of Civilian Fatalities, Number of Civilian Injured, Number of Shotguns, No Enforcement Injured,   
115                                                                                             Number of Rifles, Number of Handguns, No Enforcement Fatalities,   
58                                                                                                           No Enforcement Fatalities, Fate of Shooter Escaped,   
201                                   Number of Civilian Fatalities, Number of Handguns, No Civilian Injured, No Enforcement Fatalities, No Enforcement Injured,   
137                                                                                  No Enforcement Fatalities, No Enforcement Injured, Fate of Shooter Escaped,   
61                                                                                                              No Enforcement Injured, Fate of Shooter Escaped,   
116                                                                                                Number of Rifles, Number of Handguns, No Enforcement Injured,   
118                                                                                       Number of Rifles, No Enforcement Fatalities, Fate of Shooter Deceased,   
119                                                                                        Number of Rifles, No Enforcement Fatalities, Fate of Shooter Custody,   
208  Number of Civilian Fatalities, Number of Civilian Injured, Number of Handguns, High Civilian Casualties, No Enforcement Fatalities, No Enforcement Injured,   

In [163]:
num_officers_not_targeted = len(shooters.loc[(shooters['Number of Enforcement Fatalities'] == 0) & (shooters['Number of Enforcement Injured'] == 0)])
print("Officers were only injured or killed ", float(len(shooters) - num_officers_not_targeted) / float (len(shooters)), " of the time.")

Officers were only injured or killed  0.08358208955223881  of the time.


The market basket analysis gives us the following insights:
- Law enforcement is injured in just ~6.3% of mass shootings; officers die in 3.9% of incidents.
- Officers were neither injured nor killed in 307 out of 335 incidents (91.6%).
    - Is that discrepancy because law enforcement is adequately trained for these kinds of situations?
    - Or is it because the shooters aren't actively targeting law enforcement?
    - I'm guessing the latter, because the analysis points out that a civilian dies in 83.6% of shootings; if they were targeting law enforcement too, they'd probably have a higher 'success' rate
- 11% of the shooters had military experience

In [164]:
#Select market basket results with 'Military Experience'
military_apriori = apriori_res.loc[['Military Experience' in elem for elem in apriori_res['itemsets']]]
military_apriori

support                                              itemsets
2   0.111446                                 [Military Experience]
15  0.108434  [Number of Civilian Fatalities, Military Experience]

Looks like military experience has almost no discernable effect on the number of civilians targeted.

In [165]:
#Group the dataset by year
shooters['Date'] = pd.to_datetime(shooters['Date'])
dategroups = shooters.groupby(shooters.Date.dt.year)


#Iterate over groups, count each gun type
for name, grouped in dategroups:
    gun_counts = grouped['Type of Gun - General'].count().transpose()
    print(name)
    print(gun_counts)
    


1966
2
1971
1
1972
1
1974
2
1976
2
1979
2
1982
2
1983
2
1984
3
1985
2
1986
3
1987
1
1988
6
1989
3
1990
1
1991
5
1992
4
1993
9
1994
4
1995
4
1996
3
1997
6
1998
5
1999
7
2000
1
2001
2
2002
2
2003
3
2004
2
2005
3
2006
6
2007
10
2008
6
2009
6
2010
2
2011
6
2012
14
2013
16
2014
13
2015
65
2016
98


In [166]:
from altair import *
chart = Chart(data).mark_area(
    stacked='center',
).encode(
    color=Color('series:N',
        scale=Scale(
            range='category20b',
        ),
    ),
    x=X('date:T',
        axis=Axis(
            axisWidth=0.0,
            format='%Y',
            labelAngle=0.0,
            tickSize=0.0,
        ),
        scale=Scale(
            nice='month',
        ),
        timeUnit='yearmonth',
    ),
    y=Y('sum(count):Q',
        axis=False,
    ),
).configure_cell(
    height=200.0,
    width=300.0,
)

SchemaValidationError: Invalid specification

        altair.vegalite.v2.schema.core.MarkDef, validating 'additionalProperties'

        Additional properties are not allowed ('stacked' was unexpected)
        

In [ ]:
shooters['Fate of Shooter']